cmd/collector/main.go

```diff
@@ -25,70 +25,100 @@ import (
-	server.Register(jc.NewTChanCollectorServer(jaegerBatchesHandler))
-	server.Register(zc.NewTChanZipkinCollectorServer(zipkinSpansHandler))
+			baseMetrics := xkit.Wrap(serviceName, expvar.NewFactory(10))
+
+			builderOpts := &builder.CollectorOptions{}
+			builderOpts.InitFromViper(v)
+			sFlags := &flags.SharedFlags{}
+			sFlags.InitFromViper(v)
+			handlerBuilder, err := builder.NewSpanHandlerBuilder(
+				builderOpts,
+				sFlags,
+				basicB.Options.CassandraOption(casOptions.GetPrimary()),
+				basicB.Options.LoggerOption(logger),
+				basicB.Options.MetricsFactoryOption(baseMetrics),
+			)
+			if err != nil {
+				logger.Fatal("Unable to set up builder", zap.Error(err))
+			}
+			zipkinSpansHandler, jaegerBatchesHandler, err := handlerBuilder.BuildHandlers()
+			if err != nil {
+				logger.Fatal("Unable to build span handlers", zap.Error(err))
+			}
 
-	portStr := ":" + strconv.Itoa(*builder.CollectorPort)
-	listener, err := net.Listen("tcp", portStr)
-	if err != nil {
-		logger.Fatal("Unable to start listening on channel", zap.Error(err))
+			ch, err := tchannel.NewChannel(serviceName, &tchannel.ChannelOptions{})
+			if err != nil {
+				logger.Fatal("Unable to create new TChannel", zap.Error(err))
+			}
+			server := thrift.NewServer(ch)
+			server.Register(jc.NewTChanCollectorServer(jaegerBatchesHandler))
+			server.Register(zc.NewTChanZipkinCollectorServer(zipkinSpansHandler))
+
+			portStr := ":" + strconv.Itoa(builderOpts.CollectorPort)
+			listener, err := net.Listen("tcp", portStr)
+			if err != nil {
+				logger.Fatal("Unable to start listening on channel", zap.Error(err))
+			}
+			ch.Serve(listener)
+
+			r := mux.NewRouter()
+			apiHandler := app.NewAPIHandler(jaegerBatchesHandler, zipkinSpansHandler)
+			apiHandler.RegisterRoutes(r)
+			httpPortStr := ":" + strconv.Itoa(builderOpts.CollectorHTTPPort)
+			recoveryHandler := recoveryhandler.NewRecoveryHandler(logger, true)
+			logger.Info("Listening for HTTP traffic", zap.Int("http-port", builderOpts.CollectorHTTPPort))
+			if err := http.ListenAndServe(httpPortStr, recoveryHandler(r)); err != nil {
+				logger.Fatal("Could not launch service", zap.Error(err))
+			}
+		},
 	}
-	ch.Serve(listener)
 
-	r := mux.NewRouter()
-	apiHandler := app.NewAPIHandler(jaegerBatchesHandler, zipkinSpansHandler)
-	apiHandler.RegisterRoutes(r)
-	httpPortStr := ":" + strconv.Itoa(*builder.CollectorHTTPPort)
-	recoveryHandler := recoveryhandler.NewRecoveryHandler(logger, true)
-	logger.Info("Listening for HTTP traffic", zap.Int("http-port", *builder.CollectorHTTPPort))
-	if err := http.ListenAndServe(httpPortStr, recoveryHandler(r)); err != nil {
-		logger.Fatal("Could not launch service", zap.Error(err))
+	flagSet := &flag.FlagSet{}
+	flags.AddFlags(flagSet)
+	builder.AddFlags(flagSet)
+	casOptions.AddFlags(flagSet)
+	command.PersistentFlags().AddGoFlagSet(flagSet)
+
+	v.BindPFlags(command.PersistentFlags())
+	v.AutomaticEnv()
+	v.SetEnvKeyReplacer(strings.NewReplacer("-", "_", ".", "_"))
```

Comment:
What about an utility function which would return configured new viper instance? If so I don't know where to put it a hint would be appreciated.